### PR TITLE
Node start from epoch saving, requesting, resolving epoch start block

### PR DESCRIPTION
**Pull request by sasurobert**
Setting on bootstrap from storage the data needed for the epoch start trigger.

### CODE REVIEW

File: epochStart/metachain/trigger.go [hunk 1]
```diff
@@ -5,10 +5,15 @@ import (
...
 	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core"
 	"github.com/ElrondNetwork/elrond-go/core/check"
 	"github.com/ElrondNetwork/elrond-go/data"
 	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
 	"github.com/ElrondNetwork/elrond-go/epochStart"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/prometheus/common/log"
```
Comment (bogdan-rosianu) [inline epochStart/metachain/trigger.go]: remove this logger and use ours

File: epochStart/metachain/trigger.go [hunk 2]
```diff
@@ -156,6 +178,17 @@ func (t *trigger) SetProcessed(header data.HeaderHandler) {
 		return
 	}
 
+	metaBuff, err := t.marshalizer.Marshal(metaBlock)
+	if err != nil {
+		log.Debug("SetProcessed marshal error")
```
Comment (bogdan-rosianu) [inline epochStart/metachain/trigger.go]: `log.Debug("SetProcessed marshal", "error", err.Error())`

File: epochStart/metachain/trigger.go [hunk 3]
```diff
@@ -156,6 +178,17 @@ func (t *trigger) SetProcessed(header data.HeaderHandler) {
...
+	}
+
+	epochStartIdentifier := core.EpochStartIdentifier(metaBlock.Epoch)
+	err = t.metaHdrStorage.Put([]byte(epochStartIdentifier), metaBuff)
+	if err != nil {
+		log.Debug("SetProcessed put into metaHdrStorage error")
```
Comment (bogdan-rosianu) [inline epochStart/metachain/trigger.go]: `log.Debug("SetProcessed put into metaHdrStorage", "error", err.Error())`

File: epochStart/metachain/trigger.go [hunk 4]
```diff
@@ -177,6 +210,12 @@ func (t *trigger) Revert() {
 	t.mutTrigger.Lock()
 	defer t.mutTrigger.Unlock()
 
+	epochStartIdentifier := core.EpochStartIdentifier(t.epoch)
+	err := t.metaHdrStorage.Remove([]byte(epochStartIdentifier))
+	if err != nil {
+		log.Debug("Revert remove from metaHdrStorage error")
```
Comment (bogdan-rosianu) [inline epochStart/metachain/trigger.go]: `log.Debug("Revert remove from metaHdrStorage", "error", err.Error())`

File: epochStart/shardchain/trigger.go [hunk 5]
```diff
@@ -15,6 +15,7 @@ import (
 	"github.com/ElrondNetwork/elrond-go/marshal"
 	"github.com/ElrondNetwork/elrond-go/sharding"
 	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/prometheus/common/log"
```
Comment (bogdan-rosianu) [inline epochStart/shardchain/trigger.go]: use our logger

File: epochStart/shardchain/trigger.go [hunk 6]
```diff
@@ -236,6 +237,19 @@ func (t *trigger) updateTriggerFromMeta(metaHdr *block.MetaBlock, hdrHash []byte
 			t.epochStartRound = meta.Round
 			t.epochFinalityAttestingRound = finalityAttestingRound
 			t.epochMetaBlockHash = []byte(hash)
+
+			metaBuff, err := t.marshalizer.Marshal(meta)
+			if err != nil {
+				log.Debug("updateTriggerFromMeta marshal error")
```
Comment (bogdan-rosianu) [inline epochStart/shardchain/trigger.go]: `log.Debug("updateTriggerFromMeta marshal", "error", err.Error())`

File: epochStart/shardchain/trigger.go [hunk 7]
```diff
@@ -236,6 +237,19 @@ func (t *trigger) updateTriggerFromMeta(metaHdr *block.MetaBlock, hdrHash []byte
...
+				continue
+			}
+
+			epochStartIdentifier := core.EpochStartIdentifier(meta.Epoch)
+			err = t.metaHdrStorage.Put([]byte(epochStartIdentifier), metaBuff)
+			if err != nil {
+				log.Debug("updateTriggerMeta put into metaHdrStorage error")
```
Comment (bogdan-rosianu) [inline epochStart/shardchain/trigger.go]: `log.Debug("updateTriggerMeta put into metaHdrStorage", "error", err.Error())`

File: go.mod [hunk 8]
```diff
@@ -45,6 +45,7 @@ require (
 	github.com/pkg/errors v0.8.1
 	github.com/pkg/profile v1.3.0
 	github.com/prometheus/client_golang v1.0.0
+	github.com/prometheus/common v0.4.1
```
Comment (bogdan-rosianu) [inline go.mod]: no need for this

File: core/converters.go [hunk 9]
```diff
@@ -96,3 +97,27 @@ func SecondsToHourMinSec(input int) string {
...
 	return result
 }
+
+// EpochStartIdentifier returns the string for the epoch start identifier
+func EpochStartIdentifier(epoch uint32) string {
+	return fmt.Sprintf("epochStartBlock_%d", epoch)
+}
+
+// IsUnknownEpochIdentifier return if the epoch identifier represents unknown epoch
+func IsUnknownEpochIdentifier(identifier []byte) (bool, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/converters.go]: not really, as you have to treat separately if it is zero, or it is true. some can send bad requests and you to send the correct header, consume computational power to get the header from storage.

File: epochStart/metachain/trigger.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/metachain/trigger.go]: done

File: epochStart/metachain/trigger.go (hunk 2 above)
Comment (sasurobert) [reply epochStart/metachain/trigger.go]: done.

File: epochStart/metachain/trigger.go (hunk 3 above)
Comment (sasurobert) [reply epochStart/metachain/trigger.go]: done.

File: epochStart/metachain/trigger.go (hunk 4 above)
Comment (sasurobert) [reply epochStart/metachain/trigger.go]: done.

File: epochStart/shardchain/trigger.go (hunk 6 above)
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done

File: epochStart/shardchain/trigger.go (hunk 5 above)
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done

File: epochStart/shardchain/trigger.go (hunk 7 above)
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done

File: go.mod (hunk 8 above)
Comment (sasurobert) [reply go.mod]: done

File: epochStart/metachain/trigger.go [hunk 10]
```diff
@@ -11,11 +11,13 @@ import (
 	"github.com/ElrondNetwork/elrond-go/data/block"
 	"github.com/ElrondNetwork/elrond-go/dataRetriever"
 	"github.com/ElrondNetwork/elrond-go/epochStart"
+	"github.com/ElrondNetwork/elrond-go/logger"
 	"github.com/ElrondNetwork/elrond-go/marshal"
 	"github.com/ElrondNetwork/elrond-go/storage"
-	"github.com/prometheus/common/log"
 )
 
+var log = logger.GetOrCreate("epochStart")
```
Comment (bogdan-rosianu) [inline epochStart/metachain/trigger.go]: maybe `epochStart/metachain`. Also change for the shard trigger's logger. Right now you have the same logger key for both packages.

Comment (sasurobert) [reply epochStart/metachain/trigger.go]: this is a small  component. there is no need to declare separately for subpackages.

File: core/converters.go [hunk 11]
```diff
@@ -96,3 +97,27 @@ func SecondsToHourMinSec(input int) string {
 
 	return result
 }
+
+// EpochStartIdentifier returns the string for the epoch start identifier
+func EpochStartIdentifier(epoch uint32) string {
+	return fmt.Sprintf("epochStartBlock_%d", epoch)
```
Comment (raduchis) [inline core/converters.go]: maybe set this in a constant in core.constants and use it also in the tests if there are any

File: core/converters.go [hunk 12]
```diff
@@ -96,3 +97,27 @@ func SecondsToHourMinSec(input int) string {
...
+	if len(splitString) == 0 || len(splitString[0]) == 0 {
+		return false, ErrInvalidIdentifierForEpochStartBlockRequest
+	}
+
+	epoch, err := strconv.ParseUint(splitString[1], 10, 32)
+	if err != nil {
+		return false, ErrInvalidIdentifierForEpochStartBlockRequest
+	}
+
+	if epoch == math.MaxUint32 {
```
Comment (raduchis) [inline core/converters.go]: Why would the epoch be set to MaxUInt32?

File: epochStart/metachain/trigger.go (hunk 10 above)
Comment (raduchis) [reply epochStart/metachain/trigger.go]: yes, the logger should be separated on each package and subpackage

File: epochStart/shardchain/trigger.go [hunk 13]
```diff
@@ -12,11 +12,14 @@ import (
 	"github.com/ElrondNetwork/elrond-go/dataRetriever"
 	"github.com/ElrondNetwork/elrond-go/epochStart"
 	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/logger"
 	"github.com/ElrondNetwork/elrond-go/marshal"
 	"github.com/ElrondNetwork/elrond-go/sharding"
 	"github.com/ElrondNetwork/elrond-go/storage"
 )
 
+var log = logger.GetOrCreate("epochStart")
```
Comment (raduchis) [inline epochStart/shardchain/trigger.go]: The logger should be with "epochstart/shardchain"

File: integrationTests/sync/edgeCases_test.go [hunk 14]
```diff
@@ -20,6 +21,8 @@ func TestSyncMetaNodeIsSyncingReceivedHigherRoundBlockFromShard(t *testing.T) {
 		t.Skip("this is not a short test")
 	}
 
+	_ = logger.SetLogLevel("*:INFO,*:DEBUG,*:TRACE")
```
Comment (raduchis) [inline integrationTests/sync/edgeCases_test.go]: setLogLevel should be just one of them. This way the last one is set.

Comment (sasurobert) [reply integrationTests/sync/edgeCases_test.go]: done.

File: epochStart/shardchain/trigger.go (hunk 13 above)
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done.

File: core/converters.go (hunk 12 above)
Comment (sasurobert) [reply core/converters.go]: when running the node for the first time, you do not know what epoch the system is in.

File: core/converters.go (hunk 11 above)
Comment (sasurobert) [reply core/converters.go]: everywhere epochStartIdentifier is used, is through this method. there is not recreation in other places.

File: epochStart/metachain/trigger.go (hunk 10 above)
Comment (sasurobert) [reply epochStart/metachain/trigger.go]: done.

File: core/converters.go (hunk 12 above)
Comment (raduchis) [reply core/converters.go]: ok

File: core/converters.go (hunk 11 above)
Comment (raduchis) [reply core/converters.go]: ok

**Pull request merged**